.github/workflows/haskell-ci.yml

```diff
@@ -0,0 +1,208 @@
+        type: boolean
+        required: false
+        default: true
+      os:
+        description: |
+          Runner OS, e.g., "ubuntu-latest".
+
+          Set this to the value of `runs-on`, *not* `runner.os`.
+        type: string
+        required: true
+      sdist:
+        description: |
+          Whether to run `cabal sdist`.
+
+          This will also cause later steps to be run in the unpacked sdist.
+        type: boolean
+        required: false
+        default: true
+      test:
+        description: Whether to run `cabal test`.
+        type: boolean
+        required: false
+        default: true
+      test-targets:
+        description: Cabal targets to test, e.g. `test:my-tests`.
+        type: string
+        required: false
+        default: all
+
+# General commentary
+# ------------------
+#
+# Steps that are less likely to fail should come later, because failing earlier
+# saves time and resources. For example, `cabal check` should come after `cabal
+# build`, because edits to the Cabal file are generally more rare than edits to
+# Haskell source files.
+jobs:
+  CI:
+    runs-on: ${{ inputs.os }}
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+    - name: Install Cabal and GHC
+      uses: haskell-actions/setup@v2
+      id: setup-haskell
+      with:
+        cabal-version: ${{ inputs.cabal }}
+        ghc-version: ${{ inputs.ghc }}
+    - name: Create sdist
+      if: inputs.sdist
+      run: |
+        if [[ "${{ inputs.sdist }}" == true ]]; then
+          cabal sdist -o sdist
+          mkdir -p unpacked
+          tar -C unpacked -xzf sdist/*.tar.gz
+          cd unpacked/*
+        fi
+        echo "WORK_DIR=$(pwd)" >> $GITHUB_ENV
+    - name: Restore Cabal cache
+      uses: actions/cache/restore@v4
+      id: cache
+      env:
+        # Each `inputs.os` (e.g., `ubuntu-22.04-arm`) uniquely determines a
+        # `runner.arch` (e.g., ARM64), so there is no need to include the latter
+        # as part of the cache key.
+        key: cabal-${{ inputs.os }}-ghc${{ inputs.ghc }}
+      with:
+        path: |
+          ${{ steps.setup-haskell.outputs.cabal-store }}
+          ${{ env.WORK_DIR }}/dist-newstyle
+        # Use `github.ref` in the key for branch-specific caching.
+        #
+        # Include `github.sha` to allow saving a new cache even when there was a
+        # cache hit. See https://github.com/actions/cache/issues/109.
+        #
+        # The `restore-key` allows for restoring the cache from a different branch
+        # (e.g., the default branch) if none exists for the current branch (e.g.,
+        # on the first run on a new branch).
+        key: |
+          ${{ env.key }}-${{ github.ref }}-${{ github.sha }}
+        restore-keys: |
+          ${{ env.key }}-${{ github.ref }}
+          ${{ env.key }}-
+    - name: Configure
+      run: cabal configure ${{ inputs.configure-flags }} ${{ inputs.build-targets }} ${{ inputs.test-targets }}
```

Comment:
I think there may be a difference between supplying `all` to cabal v.s. not specifying anything.  I could be wrong, but I think not specifying anything will process all the stanzas in the main cabal file(s), whereas `all` means do all the dependencies as well.  I could be wrong, but you might want to verify this as it may impact the defaults for `inputs.build-targets` and `inputs.test-targets`.